cmd/ebrelayer/relayer/cosmos.go

```diff
@@ -94,33 +102,74 @@ func (sub CosmosSub) Start(completionEvent *sync.WaitGroup) {
 	signal.Notify(quit, syscall.SIGINT, syscall.SIGTERM)
 	defer close(quit)
 
+	var lastProcessedBlock int64
+
+	data, err := sub.DB.Get([]byte(cosmosLevelDBKey), nil)
+	if err != nil {
+		log.Println("Error getting the last cosmos block from level db", err)
+		lastProcessedBlock = 0
+	} else {
+		lastProcessedBlock = int64(binary.BigEndian.Uint64(data))
+	}
+
 	for {
 		select {
-		case result := <-out:
-			tx, ok := result.Data.(tmTypes.EventDataTx)
+		case <-quit:
+			log.Println("we receive the quit signal and exit")
+			return
+
+		case e := <-results:
+			data, ok := e.Data.(tmTypes.EventDataNewBlock)
 			if !ok {
-				sub.Logger.Error("new tx: error while extracting event data from new tx")
+				fmt.Println("we have an error")
 			}
+			blockHeight := data.Block.Height
 
-			sub.Logger.Info("New transaction witnessed")
+			// Just start from current block number if never process any block before
+			if lastProcessedBlock == 0 {
+				lastProcessedBlock = blockHeight
+			}
 
-			// Iterate over each event in the transaction
-			for _, event := range tx.Result.Events {
-				claimType := getOracleClaimType(event.GetType())
+			startBlockHeight := lastProcessedBlock + 1
+			fmt.Printf("cosmos process events from block %d to %d\n", startBlockHeight, blockHeight)
 
-				switch claimType {
-				case types.MsgBurn, types.MsgLock:
-					cosmosMsg, err := txs.BurnLockEventToCosmosMsg(claimType, event.GetAttributes())
-					if err != nil {
-						fmt.Println(err)
-						continue
+			for blockNumber := startBlockHeight; blockNumber <= blockHeight; {
+				tmpBlockNumber := blockNumber
+				block, err := client.BlockResults(&tmpBlockNumber)
+
+				if err != nil {
+					sub.Logger.Error(fmt.Sprintf("failed to get a block %s", err))
+					continue
+				}
+
+				for _, log := range block.TxsResults {
+					for _, event := range log.Events {
+
+						claimType := getOracleClaimType(event.GetType())
+
+						switch claimType {
+						case types.MsgBurn, types.MsgLock:
+							cosmosMsg, err := txs.BurnLockEventToCosmosMsg(claimType, event.GetAttributes())
+							if err != nil {
+								fmt.Println(err)
+								break
+							}
+							sub.handleBurnLockMsg(cosmosMsg, claimType)
+						}
 					}
-					// Parse event data, then package it as a ProphecyClaim and relay to the Ethereum Network
-					sub.handleBurnLockMsg(cosmosMsg, claimType)
 				}
+
+				b := make([]byte, 8)
+				binary.BigEndian.PutUint64(b, uint64(blockNumber))
```

Comment:
I don't think we need to ensure it is big endian before saving it to the db. There should not be any endian mismatch between the machine that the data is saved on and the machine it is read from.

Unless there is some leveldb semantic that makes this required, let's remove this and have it function like it does on the ethereum.go file in this directory.

Reply:
Use the big.Int for leveldb the same as ethereum.go